### ISSUE TITLE
No support for HDR gain maps (transcoded to SDR-only)

### ISSUE THREAD

**Issue by gregbenz** [libgd/libgd#898]
**Describe the bug**
When a JPG image with an HDR ("high dynamic range") gain map is transcoded, the gain map is stripped. This results in significant degradation of the image (loss of support for enhanced HDR display).

Modern displays support brighter and more vibrant HDR  images. This includes most smart phones, Apple computers, and a general trend towards widespread support on the web. More info at https://gregbenzphotography.com/hdr and https://gregbenzphotography.com/hdr-photos/jpg-hdr-gain-maps-in-adobe-camera-raw/. Instagram also supports HDR gain maps and sample images may be easily viewed with the iOS or Android app on most smart phones with comparison images at https://www.instagram.com/gregbenzphotography/

**To Reproduce**
Upload an image with a gain map to WordPress (using GD) and the resulting image in the media library will no longer no longer have a gain map.

**Expected behavior**
The image should retain a proper gain map, even after any resizing / compressing / cropping.

**Actual results**
The transcoded JPG contains only the SDR ("standard dynamic range") base image, not the required auxiliary image / metadata to support HDR display.

**Environment (please complete the following information):**
 Any version of GD will fail.

**Additional context**
The Google libultrahdr library fully supports proper transcoding of these images: https://gregbenzphotography.com/hdr-gain-map-gallery/

**Comment by cmb69** [comment]
Somewhat related to #136.

**Comment by gregbenz** [comment]
@cmb69 Certainly keeping an ICC profile is beneficial to any image (including JPG gain maps which typically have a profile in the base SDR but not typically in the auxiliary gain map image), but beyond that I'm not clear what connection you see on these issues.

_[ Note that longer term (later in 2025), AVIF images with a gain map should become a key format for sharing HDR as well. With AVIF, we can use CICP/NCLX to represent the color profile for web use. This is beneficial as CICP is 4 bytes, rather than 500-1000ish for a similar ICC profile. It also allows specifying an HDR EOTF like PQ or HLG (ICC profiles have no standard at this time for HDR). The CICP stuff is separate from the main consideration here for JPG gain maps, I'm just noting that we aren't far from a point where this more compact alternative to ICC will be a useful consideration for some file formats (a CICP tag can be added to an ICC profile as well for various reasons, but that approach would not reduce file size). ]_

**Comment by cmb69** [comment]
The connection is that both are meta-data, and libgd ignores almost all meta-data so far (heck, even Exif Orientation is ignored). That means that developers using libgd need to cater to that themselves (until libgd supports the relevant meta-data; given that the linked issue has been reported more than ten years ago, I wouldn't hold my breath).

> Upload an image with a gain map to WordPress (using GD) and the resulting image in the media library will no longer no longer have a gain map.

I suggest you report that to the WordPress developers; they might be able add support sooner than we can.

**Comment by gregbenz** [comment]
@cmb69 There are open requests here for GB, and for ImageMagick. As far as I know there is no path for WP media library support without at least one of them supporting gain maps - and ideally both (given both get a lot of use for WP).

Good to know on the metadata support, thank you.

**Comment by cmb69** [comment]
I haven't looked up where HDR gain maps are stored in JFIF files, but ICC color profiles can be handled in userland with plain PHP: get APP2 segment via `getimagesize()`, and write back using binary string manipulation. A bit fiddly, but is doable. And yes, that only caters to JFIF; but that is due to a limitation of `getimagesize()` which only reads JFIF APP markers; you could also read these (and other metadata) directly. And doing this is PHP instead of C shouldn't be a particular performance issue (contrary to image pixel manipulation).

Thinking about that, it might be best to not add support for (most) metadata to libgd, since that would drastically increase the API scope, and we're already barely focussing on image manipulation, due to the need to support yet another image format every other year.

Wrt HDR gain maps: these are to add support to image formats which don't allow to store HDR images directly. Now libgd supports AVIF and HEIF, and these formats are HDR ready; however, libgd doesn't support that, because we only support up to 8bits per color channel. If we ever extend that, it would make sense to read HDR gain maps, and apply these; and then figure out what to do with that additional information when writing back to a desired image format. However, as is it might not make much sense. Consider thumbnail creation: read JFIF including HDR gain map, resize in LDR, reapply the old HDR gain map when storing – would that even make much sense? The resizing already lost information. And if there is going to be more drastic image manipulation than resizing, keeping the old HDR gain map may even make no sense at all. This is, however, nothing libgd knows about up-front; only clients know what they are going to do with the image.

**Comment by gregbenz** [comment]
JPG gain maps store the auxiliary gain map per MPF. The Android XMP spec adds a GContainer header not required in the Adobe XMP spec. The more important ISO spec is at a draft international stage and therefore not available for review. However, it is not only supported by multiple providers, but the files are dual encoded with both Android, Adobe, or the undocumented Apple XMP - and ISO. So the ISO spec would be fully compatible with the published Adobe spec as far as the encoding of the image itself (the metadata for ISO moves from XMP to the image codestream).

**Comment by gregbenz** [comment]
libultra does not yet support AVIF / HEIF, but the team has noted that it is on their roadmap. I would think that could ultimately be managed via libultrahdr as well.

**Comment by pierrejoye** [comment]
I am planning APIs to extract/insert meta data, raw, buffer. I am not keen at add to do our own parsing/handling but providing APIs to get the meta data from any image formats, the caller can then do whatever is needed with it, reapply then when required.

Most if not all libraries specialized in such meta data accepts the data as buffer, exif, iptc, etc. So that could be helpful already and trivial then.

Some formats have been smarter and have moved some of the meta data as part of their own codecs, as they affected the loading/saving/compression/etc of their own format. That should be the way :). The rest is a nightmare if not standardized.

I understand some could be used by gd, like orientation, but where do we stop? Especially as many are not standardized at all. Open for discussions for standard cases or cases where not reading and handling them will actually affect what or how GD should process an image. Please let me know if any cases like this exists and what we could do about it. And if possible without having to implement a full meta data parser in GD, which is something I would avoid :)

**Comment by pierrejoye** [comment]
See https://github.com/libgd/libgd/pull/1002

Initial metadata extraction/insertion. No parsing, but ability to pass it to a specialized library/sdk as needed. Feedback more than welcome.

About the specific case:
"The transcoded JPG contains only the SDR ("standard dynamic range") base image, not the required auxiliary image / metadata to support HDR display."

I never used HDR before, so I am not sure if that HDR auxiliary image needs changes or not based on what the image will be modified before saving and adding the HRD meta back. F.e. resizing or similar. But I suppose that's something that could be handled of the image.

If that HDR image is also a JPEG, it may work to load it once extracted as such, or? Pointers/refs welcome if needed

**Comment by pierrejoye** [comment]
Did a quick research but enough to bring some directions.

There is little to no chance that libgd will ever support HDR natively. It is entirely a display issue. Also, logically, resizing, cropping or similar will ruin the existing HDR map. It is not resilient even if low res.

Other graphic libs, including libjpeg, completely ignore  HDR gain map

Google created libultrahdr, that wraps itself with libjpeg-turbo. This is something I could add separately, if enough projects or people need it. 

But having the current JPEG codec handle this will not be done, that's opening a can we cannot handle.

A possible solution, for a case where user do operations on the image, would to apply these ops on both the HDR gain map jpeg low res and the SDR image, then use libultrahdr to combine them back and do all the maths.

If only resize, libultra has builtin primitives for basic ops.

**Comment by gregbenz** [comment]
@pierrejoye Thank you for looking into this. I'm not sure I understand some of your comments. What do you mean by it being a "display" issue?

Resizing, cropping, compressing, rotating, and mirroring are already very well supported for JPG gain maps with libraries using libultrahdr. This includes wasm-vips, Sharp, and libvips (ImageMagick also includes it under an optional flag when compiling). You can see a wasm-vips demo here: https://gregbenzphotography.com/test/wasm-vips-hdr-demo/. This WASM library will be the basis of support for client-side transcoding in WordPress v7.1, so that takes some pressure off the need for GD to support, but it likely does not address all transcoding needs (even after media library support is added, which is not currently in the v7.1 beta, just in the editor).

I have a local demo of support via Sharp as well if that is helpful, but it is not available online as I'm not using Node JS on my server. The results there are excellent like all the others, but performance is about 2x wasm-vips due to current limitations of [libjpeg-turbo](url) in that context, so I would assume similar performance increases with GD relative to my WASM demo. 

libultrahdr is probably an ideal way to add support. Not only is it very well tested at this point, the road map will include support for AVIF gain maps (much smaller files / higher quality) and HEIF gain maps (primarily important for supporting the native gain maps created with iPhone and you'd likely transcode to JPG/AVIF gain maps with such a source).

Routing gain maps through libultrahdr has been the approach used in these other libraries (while other JPGs go through existing code paths, which. ensures maximum support for JPG overall). There is a probe feature in libultrahdr and there are other ways you could test the source.

**Comment by pierrejoye** [comment]
@gregbenz the "display issue",  I should have formulated it as "task" instead. For what I see it is mostly done by the gui layers (browser, or other apps). The hdr map is a floating point rgb image (the gray level to boost/reduce areas). The hdr data is applied most of the times by the display layer, using gpu f.e. when available, as gpu are extremely effective for such floating ops.

gd supports only 8bit per channel.. That's the biggest challenge. Vips support various floating point or custom format data as image buffer. That allows it to do hdr map operations.

Specifically for GD, adding floating point rgb(a) to the current gdImagePtr is technically possible but suboptimal. GD 3 could have it and use the more advanced surfaces and ops to do resizing, composition or similar operations. Not a small task but totally possible and cleaner than doing it in 2.x.

Alternatively, a separate set of apis using liibultrahdr would be possible too. I am not sure how watermarks or other "destructive" operations could affect the hdr composition tho'. It would allow to do whatever is needed on the sdr,  then use the ultrahdr to api to match the size/flip/etc but the altered parts of the sdr will have display disruption.

**Comment by gregbenz** [comment]
@pierrejoye got it. Should be a non-issue if libultrahdr is used, right?

**Comment by pierrejoye** [comment]
@gregbenz right. Especially as the hdr map is a lower res floating bitmap. Tests would help to see if that's a viable options.

ps: edited previous comment as an attempt to be more clear and a possible alternative option in 2.x

**Comment by gregbenz** [comment]
@pierrejoye I'd suggest limiting the support to what libultrahdr.  It can do the entire transcoding (flip, rotate, compress, resize, crop) on its own, it would only require GD to route suitable source files to that code.

If other operations such as watermarking are supported by GD, that could just route normally and be treated as the base image. Ie, not supported as HDR and the user would be responsible for creating that in the source if needed in HDR - with GD simply preserving fidelity to the source via libultrahdr.

**Comment by pierrejoye** [comment]
@gregbenz it will have to be on demand though 

I try to avoid parsing meta data as a whole library on its own would be required to do that accurately and safely. 

The flow could be:
- use existing codecs to load and process sdr, a normal joeg/heif or avif
- if a source contains hdr, done using another api using libultrahdr or iexiv2 as check first, apply the hdr map transformation to the hdr map
- pass the altered sdr and hdr map to the ultrahdr api to create the final image file


what I don't know yet, is how operations altering image contents like watermarks or the likes will look once the hdr map is applied to it. ideas?  :)

**Comment by gregbenz** [comment]
You could test for clear markers of ISO map encoding. For example, you could test JPG using:

grep -aq 'urn:iso:std:iso:ts:21496:-1' input.jpg

Those with that marker can route to libultrahdr and be handled by it. 

libultrahdr can also support the deprecated Android XMP encoding and you could test for that, but I would skip it. That content is old/limited, and your probe would probably have to test for the XMP in both base and the alt image as the XMP metadata may be altered/impartial in some edge cases, whereas the binary metadata (ISO encoding) would not be a concern for that kind of corruption.  

For content like a watermark, I'd suggest keeping that out of scope for simplicity. If you wanted to tackle it, you could apply it as you do now for the base SDR image and just set the corresponding pixels in the map to the value equal to no boost. That value depends on the range in the metadata (which is binary for ISO encoding) and may be set per channel or common for all three channels. For areas of transparency, you could weight that value with the original multiplier. It's all feasible, but this requires a number of capabilities which are either handled by libultrahdr normally or in addition to it.

**Comment by pierrejoye** [comment]
```bash
 ./Bin/uhdr tmptest/generated_uhdr.jpg tmptest/out_uhdr.jpg --resize 640x360 --rotate 90 --mirror h --sdr tmptest/out_sdr.jpg
```
```
Loaded: 1280x720 gain_map=1
Wrote UltraHDR output: tmptest/out_uhdr.jpg
Wrote SDR output: tmptest/out_sdr.jpg
```
using gd's UHDR, first attempt, easier than I thought. It supports only jpeg yet, but the enum can be extended too whatever the underlying libultrahdr will support.  Thoughts? I will do a PR later today

Validating gain map is set:

```bash
ultrahdr_app -m 1 -j tmptest/out_uhdr.jpg -o 1 -O 5 -f tmptest/gainmap_meta.txt
```
```text
--maxContentBoost 75.8243 468.126 317.151
--minContentBoost 0.0243987 0.00268363 4.95759e-05
--gamma 1 1 1
--offsetSdr 1e-07 1e-07 1e-07
--offsetHdr 1e-07 1e-07 1e-07
--hdrCapacityMin 1
--hdrCapacityMax 49.2611
--useBaseColorSpace 0
```

(using one of the libultra test image, very bright)

```
gdUhdrApplyOps
gdUhdrBytesPerPixel
gdUhdrCopyDecodedPackedImage
gdUhdrDecodeRawIntent
gdUhdrFreeCopiedRawImage
gdUhdrImageCreateFromCtx
gdUhdrImageCreateFromData
gdUhdrImageCreateFromFile
gdUhdrImageCreateFromPtr
gdUhdrImageCrop
gdUhdrImageCtx
gdUhdrImageDestroy
gdUhdrImageFile
gdUhdrImageGetSdr
gdUhdrImageHasGainMap
gdUhdrImageHeight
gdUhdrImageMirror
gdUhdrImageResize
gdUhdrImageRotate
gdUhdrImageWidth
gdUhdrImageWritePtr
gdUhdrIntAddOverflow
gdUhdrIsAvailable
gdUhdrIsSupportedFormat
gdUhdrQueueOp
gdUhdrReadAllFromCtx
gdUhdrSetError
gdUhdrUnavailableCode
gdUhdrUnavailableMessage
``

**Comment by pierrejoye** [comment]
oh, I just realized.... https://gregbenzphotography.com/hdr-gain-map-gallery/

is it you? @gregbenz  :)

**Comment by pierrejoye** [comment]
tested using the image 

<img width="800" height="534" alt="Image" src="https://github.com/user-attachments/assets/f2ee1da2-26ed-4751-8473-bc7579001887" />

from this site.

```
--maxContentBoost 75.8243 468.126 317.151
--minContentBoost 0.0243987 0.00268363 4.95759e-05
--gamma 1 1 1
--offsetSdr 1e-07 1e-07 1e-07
--offsetHdr 1e-07 1e-07 1e-07
--hdrCapacityMin 1
--hdrCapacityMax 49.2611
--useBaseColorSpace 0
```

and result image:

<img width="360" height="640" alt="Image" src="https://github.com/user-attachments/assets/318114a2-1872-427e-9fa3-e0bc04504c9a" />

the sdr:

<img width="800" height="534" alt="Image" src="https://github.com/user-attachments/assets/6c6005a0-e05f-4092-81e4-fa46279304a6" />

**Comment by gregbenz** [comment]
@pierrejoye Yes, that's me. 

You can use the attached image for testing, it is meant to expose a range of transcoding errors I have seen. You can confirm success visually using Chromium and a MacBook Pro. The image has notes on things to watch for. The map metadata should be unchanged for any simple transcode (even a crop as trying to re-optimize would invalidate a sliced image - the pieces would not match under limited headroom), only the images should change.

[ISO JPG - P3 - transcoding test - Greg Benz.jpg.zip](https://github.com/user-attachments/files/28358365/ISO.JPG.-.P3.-.transcoding.test.-.Greg.Benz.jpg.zip)

**Comment by gregbenz** [comment]
Since libultrahdr does the work and is well validated, the key areas to focus on here would be:
- suitable defaults (gain maps have additional inputs, such as the map quality - defaulting it to the same value as the base would be fine - map scaling should match source unless there is an explicit request for lower resolution - map should not drop from 3 channels to 1 without an explicit request). The wasm demo I shared above includes relevant factors to control, but we should exclude sub-sampling of the map (the map is not an image and sub-sampling invalidates it in ways which are likely to create edge cases such as noisy blue skies)
- While JPG to JPG is the only valid path now, AVIF and HEIF will be valid input / output options in the future, in case that impacts decisions now to protect for an option to request output format. I'd love to see a JXL option, but have not seen any indication this is on the roadmap.
- and of course validating that ISO-encoded JPG gain maps are routed through libultrahdr and other images route to GD's default transcoding pathway

**Comment by pierrejoye** [comment]
@gregbenz I added tests for what can be tested automatically.  Perceptual tests is not possible, so I value the meta information et.

To clarify, im this api, gd does nothing but  call libuhdr. all transformation etc api are done using libuhdr.

**Comment by gregbenz** [comment]
@pierrejoye Makes sense, just meant for a potential manual check. Took a brief look at it but didn't seem like it would build without some minor tweaks.

**Comment by pierrejoye** [comment]
@gregbenz 

"at it but didn't seem like it would build without some minor tweaks."

libuhdr or gd with libuhdr api?  

I can only test on linux and windows. Libuhdr is not yet available in most distros. if it is on Mac somehow, I can add it to the mac ci.

**Comment by gregbenz** [comment]
GD with libultra (I've built the latter on its own). I'm on Mac, so that would explain it. If it's an easy tweak for you, happy to test there.

**Comment by pierrejoye** [comment]
if you had to make changes to build on mac, please post them, I will integrate them.

Cmake uses pkgconfig to find libuhdr, so exporting the install path is required btw.

**Comment by gregbenz** [comment]
I haven't made any changes, was just going to try a quick test if it compiled.

**Comment by pierrejoye** [comment]
I see it is available with brew. Not sure it is 1.4.0. 1.4 is the minimal
version required. That's the latest version and changed apis vs previous
version. fyi :)

--
Pierre

@pierrejoye

**Comment by pierrejoye** [comment]
I added hdr to mac CI. builds and tests pass. brew has 1.4.0

**Comment by gregbenz** [comment]
Not sure I'm following, you've release a version via brew which can compile on MacOS and includes libultrahdr? I'm assuming you probably mean some other method of building.

**Comment by pierrejoye** [comment]
I mean our continuous integration runner

https://github.com/libgd/libgd/actions/runs/26612730306/job/78421897911


tests for hdr require exiv2 as well, if you run them, or see what we install in .github workflow mac yml

that's mac, with hdr enabled

**Comment by pierrejoye** [comment]
merged
